### PR TITLE
restore: CQL session created out of rpc_addresses

### LINKED ISSUE #3411: Issues when rpc_address != listen_address - Scylla Manager 3.1
Experiencing issues with their ScyllaDB setup where they have separate rpc_address and listen_address. 
Prospect is able to do CQL clients only on rpc_address, but the Scylla Manager is set up using the listen_address of a node and the cluster list works fine. However, when they try to do ```sctool restore```, it gives an error saying `"get CQL cluster session: gocql: unable to create session: unable to discover protocol version: dial tcp [<listen_address>]:9042: connect: connection refused"`. It seems that it is trying to perform some CQL operation on `listen_address`, which is not going to work.

Submit tried to update the cluster in `sctool` using `--host <rpc_address>` but then it errors out saying `"connect: connection refused - make sure the IP is correct and access to port 10001 is unblocked."` This is probably because rpc_address port 10001 isn't listening and it only listens on `listen_address`.
To address this issue, I suggest two solutions. Firstly, I recommended that prospect updates the cluster in Scylla Manager to use the `rpc_address` instead of the `listen_address`. This can be done with the `sctool` cluster update command, specifying the -host option with the rpc_address. Secondly, I suggested that the prospect changes the Scylla configuration on the node to listen for CQL connections on the `listen_address`. This would involve changing the `rpc_address` at scylla.yaml file to the `listen_address`, and then restart the Scylla service.
However, tried both solutions, but they failed.

```
# sctool status

╭────┬──────────┬──────────┬────────────────────────────┬─────────────┬──────┬─────────┬────────┬───────┬──────────────────────────────────────╮
│UN  │ CQL      │ REST     │    Address                 │ Uptime      │ CPUs │ Memory  │ Scylla │ Agent │ Host ID                              │
├────┼──────────┼──────────┼────────────────────────────┼─────────────┼──────┼─────────┼────────┼───────┼──────────────────────────────────────┤
│ UN │ UP (8ms) │ UP (0ms) │ xxxx:xxxx:x:x:xx:xxx:664:11 │ 2360h11m51s │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ 080acc55-8710-4502-9ea3-0ef30689d8d4 │
│ UN │ UP (8ms) │ UP (1ms) │ xxxx:xxxx:x:x:xx:xxx:664:12 │ 2359h28m5s  │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ bafe23f9-1228-493f-b0e1-f13c8db1a1be │
│ UN │ UP (7ms) │ UP (0ms) │ xxxx:xxxx:x:x:xx:xxx:664:13 │ 213h6m1s    │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ 95d7086f-eaac-417d-9db9-153e5e0c09ac │
│ UN │ UP (9ms) │ UP (0ms) │ xxxx:xxxx:x:x:xx:xxx:664:14 │ 213h18m41s  │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ ea407b4e-9f5e-436d-b349-daabe11cde1f │
│ UN │ UP (8ms) │ UP (5ms) │ xxxx:xxxx:x:x:xx:xxx:664:15 │ 2162h8m11s  │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ b3e81033-cee1-48fa-94a9-7074a7d15979 │
╰────┴──────────┴──────────┴────────────────────────────┴─────────────┴──────┴─────────┴────────┴───────┴──────────────────────────────────────╯
root@stg-scylladbmanager ~ Thu Jun 01 09:51:50
#
root@stg-scylladbmanager ~ Thu Jun 01 09:51:50
#
root@stg-scylladbmanager ~ Thu Jun 01 09:51:50
# telnet xxxx:xxxx:x:x:xx:xxx:664:11 9042
Trying 2a0b:b582::10:123:64:11...
telnet: connect to address 2a0b:b582::10:123:64:11: Connection refused
root@stg-scylladbmanager ~ Thu Jun 01 09:51:52
# telnet xxxx:xxxx:x:x:xx:xxx:648:11 9042
Trying xxxx:xxxx::xx:xxx:48:11...
Connected to xxxx:xxxx:x:x:xx:xxx:648:11.
Escape character is '^]'.
^CConnection closed by foreign host.# telnet xxxx:xxxx:x:x:xx:xxx:664:11 10001
Trying 2a0b:b582::10:123:64:11...
Connected to xxxx:xxxx:x:x:xx:xxx:664:11.
Escape character is '^]'.
^CConnection closed by foreign host.
root@stg-scylladbmanager ~ Thu Jun 01 09:54:59
# telnet xxxx:xxxx:x:x:xx:xxx:664:48 10001
Trying 2a0b:b582::10:123:64:48...
telnet: connect to address 2a0b:b582::10:123:64:48: No route to host


# cat /etc/scylla/scylla.yaml | grep -e "listen_address\|rpc_address" | grep -v "#"
listen_address: 2a0b:b582::10:123:64:11
rpc_address: xxxx:xxxx::xx:xxx:48:11# firewall-cmd --list-all
public (active)
target: default
icmp-block-inversion: no
interfaces: ens192 eth1 eth2 eth3
sources:
services: dhcpv6-client ssh zabbix-agent
ports: 161/udp 9042/tcp 7001/tcp 7199/tcp 123/udp 61621/tcp 10001/tcp 9180/tcp 3000/tcp 9100/tcp 1514/tcp 19042/tcp 10050/tcp
protocols:
forward: no
masquerade: yes
forward-ports:
source-ports:
icmp-blocks:
rich rules:# telnet 2a0b:b582::10:123:64:11 9042
Trying 2a0b:b582::10:123:64:11...
telnet: connect to address 2a0b:b582::10:123:64:11: Connection refusedroot@stg-scylladb101 ~ Thu Jun 01 09:58:30
# telnet 2a0b:b582::10:123:64:11 10001
Trying 2a0b:b582::10:123:64:11...
Connected to 2a0b:b582::10:123:64:11.
Escape character is '^]'.
^CConnection closed by foreign host.root@stg-scylladb101 ~ Thu Jun 01 09:58:36
# telnet xxxx:xxxx::xx:xxx:48:11 9042
Trying xxxx:xxxx::xx:xxx:48:11...
Connected to xxxx:xxxx::xx:xxx:48:11.
Escape character is '^]'.
^CConnection closed by foreign host.root@stg-scylladb101 ~ Thu Jun 01 09:58:48
# telnet xxxx:xxxx::xx:xxx:48:11 10001
Trying xxxx:xxxx::xx:xxx:48:11...
telnet: connect to address xxxx:xxxx::xx:xxx:48:11: Connection refused
```

1. Setting(adding) listen_address in: scylla-manager-agent.yaml
```Jun 05 12:04:11 stg-scylladb101 systemd[1]: Started Scylla Manager Agent.
Jun 05 12:04:11 stg-scylladb101 scylla-manager-agent[3619903]: STARTUP ERROR: yaml: unmarshal errors:
Jun 05 12:04:11 stg-scylladb101 scylla-manager-agent[3619903]:   line 2: field listen_address not found in type agent.Config
```
2. Set https: xxxx:xxxx::xx:xxx:48:11:10001 [same as rpc_address] in: scylla-manager-agent.yaml
```Jun 05 11:56:31 stg-scylladb101 scylla-manager-agent[3619686]: {"L":"ERROR","T":"2023-06-05T11:56:31.291Z","M":"Bye","error":"HTTPS server start: listen tcp: address 2a0b:b582::10:123:48:1>
Jun 05 11:56:31 stg-scylladb101 scylla-manager-agent[3619686]: STARTUP ERROR: HTTPS server start: listen tcp: address xxxx:xxxx::xx:xxx:48:11:10001: too many colons in address
```
3. Set https: 0.0.0.0:10001 [ to listen to all address both rpc and listen_address] in: scylla-manager-agent.yamlAgent works fine, listens to both Set xxxx:xxxx::xx:xxx:48:11:10001 and 2a0b:b582::10:123:64:11:10001
Scylla-Manager nodes can also reach it(verified via telnet).
```# telnet xxxx:xxxx:x:x:xx:xxx:648:11 10001
Trying xxxx:xxxx::xx:xxx:48:11...
Connected to xxxx:xxxx:x:x:xx:xxx:648:11.
Escape character is '^]'.
```
Updated Cluster:
```
# sctool cluster update -c "ClusterName" --host xxxx:xxxx:x:x:xx:xxx:648:11
root@stg-scylladbmanager ~ Mon Jun 05 11:34:18
```
Yet it shows IP address of subnet 64:
```
╭────┬───────────┬──────────┬────────────────────────────┬─────────────┬──────┬─────────┬────────┬───────┬──────────────────────────────────────╮
│    │ CQL       │ REST     │ Address                    │ Uptime      │ CPUs │ Memory  │ Scylla │ Agent │ Host ID                              │
├────┼───────────┼──────────┼────────────────────────────┼─────────────┼──────┼─────────┼────────┼───────┼──────────────────────────────────────┤
│ UN │ UP (9ms)  │ UP (3ms) │ xxxx:xxxx:x:x:xx:xxx:664:11 │ 2457h57m35s │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ 080acc55-8710-4502-9ea3-0ef30689d8d4 │
│ UN │ UP (10ms) │ UP (4ms) │ xxxx:xxxx:x:x:xx:xxx:664:12 │ 2457h11m19s │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ bafe23f9-1228-493f-b0e1-f13c8db1a1be │
│ UN │ UP (10ms) │ UP (0ms) │ xxxx:xxxx:x:x:xx:xxx:664:13 │ 310h51m15s  │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ 95d7086f-eaac-417d-9db9-153e5e0c09ac │
│ UN │ UP (9ms)  │ UP (2ms) │ xxxx:xxxx:x:x:xx:xxx:664:14 │ 311h1m40s   │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ ea407b4e-9f5e-436d-b349-daabe11cde1f │
│ UN │ UP (7ms)  │ UP (2ms) │ xxxx:xxxx:x:x:xx:xxx:664:15 │ 2259h51m10s │ 4    │ 31.180G │ 5.1.5  │ 3.1.0 │ b3e81033-cee1-48fa-94a9-7074a7d15979 │
╰────┴───────────┴──────────┴────────────────────────────┴─────────────┴──────┴─────────┴────────┴───────┴──────────────────────────────────────╯
```
And restore fails:
```
# sctool restore  --cluster "ClusterName"  -L 's3:scylla-backup01-stg-20bda0ad' --
num-retries 1 -T sm_20230531105239UTC -K cycling.cyclist_base --restore-tables --dry-run --show-tables
Error: get restore target: list all views of cluster 4e81165b-7057-4de6-b2bf-2e553c684ad7: get CQL cluster session: gocql: unable to create session: unable to discover protocol version: dial tcp [xxxx:xxxx::xx:xxx:64:14]:9042: connect: connection refused
Trace ID: wxOQApYdSyWTPb7mXiJNxQ (grep in scylla-manager logs)
```

[Archive.zip](https://github.com/scylladb/scylla-manager/files/11673178/Archive.zip)

### PR DESCRIPTION
**Pull request by karol-kokoszka**
Fixes #3411 

This PR extends docker-compose.yaml used for integration-tests to include additional network.
Scylla's `listen_address` belongs to `second` network.
Scylla's `rpc_address` belongs to `public` network.

Scylla uses `rpc_address` to accept CQL connections when `listen_address` is used for node2node communication.

Issue #3411 showed that restore procedure in scylla-manager doesn't support such a setup.
This PR addresses that by querying ScyllaAPI to get the RPC address when creating CQL session.

Integration-tests are updated to work fine on current docer-compose environment.

`db.go` from `testutils` is moved to another package `testutils/db` to avoid import cycle.

---

"Only the Best is Good Enough." - LEGO company motto

Please make sure that:
- Code is split to commits that address a single change
- Commit messages are informative
- Commit titles have module prefix
- Commit titles have issue nr. suffix

### CODE REVIEW

File: pkg/service/cluster/service.go [hunk 1]
```diff
@@ -604,3 +602,18 @@ func (s *Service) notifyChangeListener(ctx context.Context, c Change) error {
...
+	for _, h := range hosts {
+		ni, err := client.NodeInfo(ctx, h)
+		if err != nil {
+			combinedError = multierr.Append(combinedError, err)
+			continue
+		}
+		sessionHosts = append(sessionHosts, ni.CQLAddr(h))
+	}
+
+	return sessionHosts, combinedError
```
Comment (Michal-Leszczynski) [inline pkg/service/cluster/service.go:621]: It's often the case that the length of returned addresses is checked:
```go
sessionHosts, err := GetRPCAddresses(ctx, client, client.Config().Hosts)
	if len(sessionHosts) == 0 {
		return session, errors.Wrapf(err, "failed to build list of session ips")
	}
```
Maybe it's better to move it inside the `GetRPCAddresses`?

**Comment by Michal-Leszczynski** [comment]
I understand that this issue has been found while using restore, but changes introduced in this PR do not change it's code anywhere. So is it really a restore issue?

Shouldn't commit `restore: CQL session created out of rpc_addresses` have `cluster` prefix?

File: pkg/service/cluster/service.go [hunk 2]
```diff
@@ -527,8 +527,8 @@ func (s *Service) GetSession(ctx context.Context, clusterID uuid.UUID) (session
 	}
 
 	sessionHosts, err := GetRPCAddresses(ctx, client, client.Config().Hosts)
-	if len(sessionHosts) == 0 {
-		return session, errors.Wrapf(err, "failed to build list of session ips")
+	if err != nil && len(sessionHosts) == 0 {
+		return session, errors.Wrap(err, "fetch RPC addresses")
```
[the remark this answers is not in the document]
Comment (karol-kokoszka) [reply pkg/service/cluster/service.go]: yes, they should, will fix

File: pkg/service/cluster/service.go [hunk 3]
```diff
@@ -603,6 +603,9 @@ func (s *Service) Close() {
 	s.clientCache.Close()
 }
 
+// GetRPCAddresses accepts client and hosts parameters that are used later on to query client.NodeInfo endpoint
+// returning RPC addresses for given hosts.
+// RPC addresses are the ones that scylla uses to accept CQL connections.
```
[the remark this answers is not in the document]
Comment (karol-kokoszka) [reply pkg/service/cluster/service.go:608]: they should, thanks

File: pkg/service/cluster/service.go (hunk 1 above)
Comment (karol-kokoszka) [reply pkg/service/cluster/service.go:621]: actually makes sesne

**Comment by karol-kokoszka** [comment]
@Michal-Leszczynski it's not really the restore issue, but restore is affected the most.

File: pkg/service/cluster/service.go (hunk 1 above)
Comment (karol-kokoszka) [reply pkg/service/cluster/service.go:621]: ok, I cannot move it inside GetRPCAddresses.
GetRPCAddresses returns combined errors from every client.NodeInfo invocation. 
I don't want to extend the function signature to pass logger here and log this information in the method itself.

What I did instead is I created ErrNoRPCAddressesFound which is returned (together with other ones) when there is no address found.
Then it's up to the end user to check this error and react.

File: pkg/ping/dynamoping/main_test.go [hunk 4]
```diff
@@ -1,13 +0,0 @@
...
-
-import (
-	"testing"
-
-	"go.uber.org/goleak"
-)
-
-func TestMain(m *testing.M) {
-	goleak.VerifyTestMain(m)
-}
```
[the remark this answers is not in the document]
Comment (karol-kokoszka) [reply pkg/ping/dynamoping/main_test.go:13]: https://github.com/census-instrumentation/opencensus-go/issues/1191

[the remark this answers is not in the document]
Comment (karol-kokoszka) [reply pkg/ping/dynamoping/main_test.go]: The question about the reason caused me to double check the goleak output....
The one from dynamoping is caused by opencensus added to google api. I just disabled the check directly in main_test.

But, there was another leak in CQLPing.
It started to appear in this PR as the `cqlping_integration_test.go/TestPingIntegration` is extended to use scyllaclient to query RPC addresses.
Leak from CQLPing was not connected to opencensus only... Seems we have had leaking goroutines in scyllaclient. We haven't been closing custom host pool used by the transporn in scyllaclient.

So.... TestMains are bring back and there is additional fix in scyllaclient to close the host pool.

[the remark this answers is not in the document]
Comment (karol-kokoszka) [reply pkg/ping/dynamoping/main_test.go]: The fix is in this commit https://github.com/scylladb/scylla-manager/pull/3418/commits/b99241cfbdcb44a88d0ab786ee785ecf3f9d0ead

**Pull request merged**